File: Data Structures & Algorithms/longest-common-subsequence/submission-3.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        # Create memo table for each possibility of text1 and text2 positions
        memo = [len(text2) * [-1] for _ in range(len(text1))]

        # Recursively go through and compare each position of the texts
        def dfs(i, j):
            # Check if indices are inbounds
            if i >= len(text1) or j >= len(text2):
                return 0
            # If texts are equal, set memo table and continue to recurse
            if text1[i] == text2[j]:
                memo[i][j] = 1 + dfs(i + 1, j + 1)
                return memo[i][j]
            # If text are not equal, check if memo table has a value for the current positions
            if memo[i][j] != -1:
                return memo[i][j]
            
            # Otherwise, set the current position's value in the memo table equal to the
            # max value of the recursion between incrementing text1's pointer and leaving text2's
            # pointer the same and vice-versa
            memo[i][j] = max(dfs(i+1, j), dfs(i, j+1))
            # Return the value of the current position
            return memo[i][j]
        
        return dfs(0, 0)

File: Data Structures & Algorithms/interleaving-string/submission-3.java
class Solution {
    private String s1, s2, s3;
    private int n1, n2, n3;
    private Boolean[][] memo;

    public boolean isInterleave(String s1, String s2, String s3) {
        this.n1 = s1.length();
        this.n2 = s2.length();
        this.n3 = s3.length();

        this.s1 = s1;
        this.s2 = s2;
        this.s3 = s3;

        this.memo = new Boolean[n1 + 1][n2 + 1];

        if (n1 + n2 != n3)
        {
            return false;
        }

        return dfs(0, 0);
    }

    public boolean dfs(int i, int j)
    {
        if (i == this.n1 && j == this.n2 && i + j == this.n3)
        {
            return true;
        }

        if (this.memo[i][j] != null)
        {
            return this.memo[i][j];
        }

        boolean match_s1 = false;
        boolean match_s2 = false;
        if (i < n1 && this.s1.charAt(i) == this.s3.charAt(i + j))
        {
            match_s1 = dfs(i + 1, j);
        }
        if (j < n2 && this.s2.charAt(j) == this.s3.charAt(i + j))
        {
            match_s2 = dfs(i, j + 1);
        }

        this.memo[i][j] = match_s1 || match_s2;
        return this.memo[i][j];
    }
}
